src/dftbp/dftb/elecconstraints.F90

```diff
@@ -7,428 +7,396 @@
-    complex(dp), allocatable :: Lz(:,:)
-    complex(dp), allocatable :: Lplus(:,:)
-    real(dp), allocatable :: tmpBlock(:,:)
-    real(dp) :: Lshell(3), Sshell(3), W, vecNorm(3)
+    !> Energy
+    real(dp), intent(in) :: energy
 
-    complex(dp), parameter :: i = (0.0_dp,1.0_dp)
+    !> Contribution to free energy functional from constraint(s)
+    real(dp), intent(out) :: deltaW
 
-    nSpin = size(shift,dim=4)
+    !> Maximum derivative of energy functional with respect to Vc
+    real(dp), intent(out) :: dWdVcMax
 
-    vecNorm = vec / sqrt(sum(vec**2))
+    !> Gradient convergence achieved
+    logical, intent(out) :: tConverged
 
-    allocate(SpeciesL(orb%mOrb,orb%mOrb,3))
-    SpeciesL = 0.0_dp
-    allocate(Lz(orb%mOrb,orb%mOrb))
-    allocate(Lplus(orb%mOrb,orb%mOrb))
+    !! Derivative of energy functional with respect to Vc
+    real(dp), allocatable :: dWdVc(:)
 
-    iSp = species(conAt)
-    Lz = 0.0_dp
-    Lplus = 0.0_dp
-    iSh = orb%angShell(conSh,iSp)
-    call getLoperators(iSh, Lplus(1:2*iSh+1,1:2*iSh+1),Lz(1:2*iSh+1,1:2*iSh+1))
-    speciesL(orb%posShell(conSh,iSp):orb%posShell(conSh+1,iSp)-1,orb%posShell(conSh,iSp): &
-        & orb%posShell(conSh+1,iSp)-1,1) = aimag(Lplus(1:2*iSh+1,1:2*iSh+1))
-    speciesL(orb%posShell(conSh,iSp):orb%posShell(conSh+1,iSp)-1,orb%posShell(conSh,iSp): &
-        & orb%posShell(conSh+1,iSp)-1,2) = -real(Lplus(1:2*iSh+1,1:2*iSh+1))
-    speciesL(orb%posShell(conSh,iSp):orb%posShell(conSh+1,iSp)-1,orb%posShell(conSh,iSp): &
-        & orb%posShell(conSh+1,iSp)-1,3) = aimag(Lz(1:2*iSh+1,1:2*iSh+1))
+    !! Potential displacement proposed by optimizer
+    real(dp) :: potDisplace(size(this%Vc))
 
-    allocate(tmpBlock(orb%mOrb,orb%mOrb))
+    !! Iterates over constraints
+    integer :: iConstr
 
-    Lshell = 0.0_dp
+    !! Number of constraints requested by the user
+    integer :: nConstr
 
-    iOrb = orb%nOrbSpecies(iSp)
-    tmpBlock(:,:) = 0.0_dp
+    nConstr = size(this%wAt)
+    allocate(dWdVc(nConstr))
+    dWdVc(:) = 0.0_dp
+    deltaW = 0.0_dp
 
-    ! identity part
-    tmpBlock(1:iOrb,1:iOrb) = qBlockSkew(1:iOrb,1:iOrb,conAt,1)
-    iStart = orb%posShell(conSh,iSp)
-    iEnd = orb%posShell(conSh+1,iSp)-1
-    do ii = 1, 3
-      Lshell(ii) = -sum(SpeciesL(iStart:iEnd,iStart:iEnd,ii) &
-          & * transpose(tmpBlock(iStart:iEnd,iStart:iEnd)))
+    do iConstr = 1, nConstr
+      call getConstrainEnergyAndPotQ(deltaW, dWdVc(iConstr), this%Vc(iConstr), this%Nc(iConstr),&
+          & this%wAt(iConstr)%data, this%wAtOrb(iConstr)%data, this%wAtSpin(iConstr)%data,&
+          & qq)
     end do
 
-    Sshell = sum(qIn(orb%posShell(conSh,species(conAt)): &
-        & orb%posShell(conSh+1,species(conAt))-1,conAt,2:4),dim=1)
+    ! Optimizers set up to minimize, therefore sign change in total energy and gradients
+    ! call next(this%potOpt, -(energy + deltaW), -dWdVc, this%Vc, tConverged)
+    call this%potOpt%step(energy + deltaW, -dWdVc, potDisplace)
```

Comment:
I find the comment (and the commented out code) confusing, as it is the opposite (for the energy) as actually done in the line below.

Reply:
Artifact of old version, removed in latest iteration.